lib/analyzers/two_fer/analyze.rb

```diff
@@ -0,0 +1,201 @@
+module TwoFer
+  class Analyze < ExerciseAnalyzer
+    include Mandate
+
+    def analyze!
+      ####
+      # First we check that there is a two-fer class or module
+      # and that it contains a method called two-fer
+      two_fer_module = ExtractModuleOrClass.(root_node, "TwoFer")
+      return fail_with_no_module unless two_fer_module
+
+      two_fer_method = ExtractClassMethod.(two_fer_module, "two_fer")
+      return fail_with_no_method unless two_fer_method
```

Comment:
I think there are lots of good reasons to do this.

`ExtractClassMethod` has two methods:
```ruby
  def on_sclass(node)
    # If we're in a class << self context
    if node.children[0].self_type?
      @found_method = ExtractInstanceMethod.(node.children[1], name)
    end
  end

  def on_defs(node)
    if node.method_name == name
      @found_method = node
    end
  end
```

and `ExtractInstanceMethod` which it potentially calls has
```ruby
  def on_def(node)
    if node.method_name == name
      @found_method = node
    end
  end
```

I can't imagine a solution where we wouldn't want to quickly find classes/modules or methods in the codebase to analyze, so I'm not sure why having this duplication in every analyzer is valuable? Having a helper class that extracts out a relevant node to work on seems like a sensible approach to me, and more less complicated than overloading `on_def` for multiple purposes in this analyzer.

Reply:
Ah, I see where you're coming from.

I'd like to avoid all extra abstractions until we actually know what the code looks like. I find that it makes it much harder to think about things if we've added abstractions up front. (It's often fine when writing code that is similar to code we've written a bajillion times before, but here the domain is new, so I'd like to just have duplication until we actually understand what we like and what we hate).

Reply:
Yeah. Makes sense. 

Reflecting more, I think the key reason I started doing this was that the static analysis hurts my head to visualize the flow of, so extracting things into clear abstractions (rather than having inline `on_class`, `on_method`) made it clear for me to think about. The abstraction for reuse was probably an afterthought to the abstraction for readability.